### PR TITLE
Specified conventions for core packages structure and namespaces

### PR DESCRIPTION
**Pull request by alongosz**
| Question      | Answer
| ------------- | ---
| JIRA Ticket   | N/A
| Versions      | 3.1+

This PR outlines know-how of eZ Platform by Ibexa core packages structure and namespace naming conventions.

This specification will be available from the side menu of the Documentation: `Resources and community > Contributing > Package structure specification`. Is it OK, or is a different place preferred?

### Open questions:
- [x] test namespace: I see that already existing 3.1 implementations utilize the pattern `Ibexa\Platform\<PackageName>\Tests\` mapped to `tests/lib` and similar one for bundles. 
  Given that `Ibexa\Platform\<PackageName>` is mapped to `src/lib`, this creates an overlapping conflict. It's not possible to create a subdirectory `src/lib/Tests` because it's gonna be expected in `tests/lib`. 
  While a need for this directory is definitly some distant edge case (e.g AB-Testing framework impl.?), it's a matter of consistency. // using `Ibexa\Platform\Tests\<PackageName>`
  Is this a good choice or can it be changed the way this PR ATM describes - using generic `Ibexa\Platform\Tests` prefix?
- [x] Bundle class name. I see that some existing implementations follow the pattern `Platform<PackageName>Bundle` instead of `<PackageName>Bundle`. Does it make sense to repeat "Platform"? // yes, because: https://github.com/ezsystems/developer-documentation/pull/1094#discussion_r448280460

### CODE REVIEW

File: docs/community_resources/package_structure.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+Example:
+
+```php
+namespace Ibexa\Platform\Search;
+```
+
+A package which groups some DXP features MAY use an additional prefix, for example:
+
+```php
+namespace Ibexa\Platform\Commerce\Shop;
```
Comment (ViniTou) [inline docs/community_resources/package_structure.md:53]:
```suggestion
namespace Ibexa\Commerce\Shop;
```
@ezsystems/php-dev-team

File: docs/community_resources/package_structure.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+```
+
+### Bundles
+
+The bundle class definition in the **`src/bundle`** directory MUST BE as follows:
+
+```php
+namespace Ibexa\Platform\Bundle\<PackageName>Bundle;
+
+class <PackageName>Bundle // ...
```
Comment (ViniTou) [inline docs/community_resources/package_structure.md:63]:
```suggestion
class Ibexa<ProductName><PackageName>Bundle // ...
```

File: docs/community_resources/package_structure.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+```
+
+Example:
+```php
+namespace Ibexa\Platform\Bundle\SearchBundle;
+
+class SearchBundle // ...
```
Comment (ViniTou) [inline docs/community_resources/package_structure.md:70]:
```suggestion
class IbexaPlatformSearchBundle // ...
```

As that name is also used in twig templates, and using just `@Search` could lead to some aliases problem, and it is not clear at first glance from where given resource comes from.

Comment (alongosz) [reply docs/community_resources/package_structure.md:70]: > As that name is also used in twig templates, and using just `@Search` could lead to some aliases problem, and it is not clear at first glance from where given resource comes from.

This is very good point, thanks!

File: docs/community_resources/package_structure.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+This specification **IS NOT meant** for 3rd party packages or projects.
+
+What needs to be emphasised is that the conventions described here are meant to be applied as an evolution.
+It means that while the existing code won't be changed at once (or at all in some cases), the new code
+needs to follow the rules outlined here.
+
+## Root PHP namespace
+
+eZ Platform by Ibexa core PHP code needs to be defined in a namespace with the following prefix:
```
Comment (alongosz) [inline docs/community_resources/package_structure.md:13]:
```suggestion
Ibexa core PHP code needs to be defined in a namespace with the following prefix:
```

File: docs/community_resources/package_structure.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+```php
+namespace Ibexa\Platform;
```
Comment (alongosz) [inline docs/community_resources/package_structure.md:16]:
```suggestion
namespace Ibexa;
```

File: docs/community_resources/package_structure.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+```
+is meant for internal implementation not tied to the Symfony Framework.
+
+Example:
+
+```php
+namespace Ibexa\Platform\Search;
+```
+
+A package which groups some DXP features MAY use an additional prefix, for example:
```
Comment (alongosz) [inline docs/community_resources/package_structure.md:50]: Already mentioned before, so this sentence is redundant:
```suggestion
```

File: docs/community_resources/package_structure.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+needs to follow the rules outlined here.
+
+## Root PHP namespace
+
+eZ Platform by Ibexa core PHP code needs to be defined in a namespace with the following prefix:
+
+```php
+namespace Ibexa\Platform;
+```
+
```
Comment (alongosz) [inline docs/community_resources/package_structure.md:18]:
```suggestion

Additionally each Product from Ibexa DXP should define its own prefix, e.g:
* `namespace Ibexa\Platform;`
* `namespace Ibexa\Commerce;`
* `namespace Ibexa\Personalization;`

```

**Comment by alongosz** [comment]
Ready for English polishing (pun intended) review @DominikaK :tada:

File: docs/community_resources/package_structure.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+This specification **IS NOT meant** for 3rd party packages or projects.
+
+What needs to be emphasised is that the conventions described here are meant to be applied as an evolution.
+It means that while the existing code won't be changed at once (or at all in some cases), the new code
+needs to follow the rules outlined here.
+
+## Root PHP namespace
+
+eZ Platform by Ibexa core PHP code needs to be defined in a namespace with the following prefix:
```
Comment (DominikaK) [inline docs/community_resources/package_structure.md:13]:
```suggestion
Define eZ Platform by Ibexa core PHP code in a namespace with the following prefix:
```

File: docs/community_resources/package_structure.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+```php
+namespace Ibexa\Platform;
+```
+
+A package which groups some DXP features MAY use an additional prefix, for example:
```
Comment (DominikaK) [inline docs/community_resources/package_structure.md:19]:
```suggestion
A package which groups some DXP features can use an additional prefix, for example:
```

File: docs/community_resources/package_structure.md [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+namespace Ibexa\Platform\Commerce;
+```
+
+```php
+namespace Ibexa\Platform\Personalization;
+```
+
+## Packages
+
+The general package directory structure and corresponding PHP namespace mapping is as follows.
```
Comment (DominikaK) [inline docs/community_resources/package_structure.md:31]:
```suggestion
The general package directory structure and corresponding PHP namespace mapping are:

```

File: docs/community_resources/package_structure.md [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+|   +-- bundle (`Ibexa\Platform\Bundle\<PackageName>`)
+|   +-- contracts (`Ibexa\Platform\Contracts\<PackageName>`)
+|   +-- lib (`Ibexa\Platform\<PackageName>`)
++-- tests
+|   +-- bundle (`Ibexa\Platform\Tests\Bundle\<PackageName>`)
+|   +-- integration (`Ibexa\Platform\Tests\Integration\<PackageName>`)
+|   +-- lib (`Ibexa\Platform\Tests\<PackageName>`)
+```
+
+If a package does not contain some of the described parts, those directories MAY BE skipped.
```
Comment (DominikaK) [inline docs/community_resources/package_structure.md:44]:
```suggestion
If a package does not contain some of the described parts, you can skip those directories.
```

File: docs/community_resources/package_structure.md [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+
+### Implementation (lib)
+
+The **`src/lib`** directory and its corresponding namespace:
```
Comment (DominikaK) [inline docs/community_resources/package_structure.md:48]:
```suggestion
The `src/lib` directory and its corresponding namespace are:

```

File: docs/community_resources/package_structure.md [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+```php
+namespace Ibexa\Platform\<PackageName>;
+```
+is meant for internal implementation not tied to the Symfony Framework.
```
Comment (DominikaK) [inline docs/community_resources/package_structure.md:52]:
```suggestion

This directory is meant for internal implementation not tied to the Symfony Framework.

```

File: docs/community_resources/package_structure.md [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+namespace Ibexa\Platform\Search;
+```
+
+```php
+namespace Ibexa\Platform\Commerce\Shop;
+```
+
+### Bundles
+
+The bundle class definition in the **`src/bundle`** directory MUST BE as follows:
```
Comment (DominikaK) [inline docs/community_resources/package_structure.md:66]:
```suggestion
The bundle class definition in the `src/bundle` directory must be:
```

File: docs/community_resources/package_structure.md [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+
+```php
+namespace Ibexa\Platform\Bundle\<PackageName>;
+
+class IbexaPlatform[ProductGroup]<PackageName>Bundle // ...
+```
+
+Examples:
```
Comment (DominikaK) [inline docs/community_resources/package_structure.md:74]:
```suggestion
Examples:

```

File: docs/community_resources/package_structure.md [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+
+```php
+namespace Ibexa\Platform\Bundle\Commerce\Shop;
+
+class IbexaPlatformCommerceShopBundle // ...
+```
+
+### Contracts
+
+A package MAY introduce a namespace for contracts, to be consumed by 1st and 3rd party packages
```
Comment (DominikaK) [inline docs/community_resources/package_structure.md:89]:
```suggestion
A package may introduce a namespace for contracts, to be consumed by first and third party packages
```

File: docs/community_resources/package_structure.md [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+and projects, which MUST be prefixed as:
```
Comment (DominikaK) [inline docs/community_resources/package_structure.md:90]:
```suggestion
and projects, which must be prefixed as:
```

File: docs/community_resources/package_structure.md [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+
+```php
+namespace Ibexa\Platform\Contracts\SiteFactory;
+```
+
+```php
+namespace Ibexa\Platform\Contracts\Commerce\Shop;
+```
+
+That namespace needs to be mapped to the **`src/contracts`** directory of a package.
```
Comment (DominikaK) [inline docs/community_resources/package_structure.md:110]:
```suggestion
That namespace needs to be mapped to the `src/contracts` directory of a package.

```

File: docs/community_resources/package_structure.md (hunk 13 above)
Comment (alongosz) [inline docs/community_resources/package_structure.md:52]: Rephrased after internal sync:
```suggestion
The `src/lib` directory and its corresponding `Ibexa\Platform\<PackageName>` namespace are meant for internal implementation not tied to the Symfony Framework.
```
